File: Data Structures & Algorithms/find-target-in-rotated-sorted-array/submission-1.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        l = 0
        r = len(nums) - 1

        while l <= r:
            m = (l + r) // 2

            if target == nums[m]:
                return m

            # Left sorted portion
            if nums[m] >= nums[l]:
                if nums[m] < target or nums[l] > target:
                    l = m + 1
                else:
                    r = m - 1
            # Right Sorted portion
            else:
                if nums[m] > target or nums[r] < target:
                    r = m - 1
                else:
                    l = m + 1

        return -1

File: Data Structures & Algorithms/products-of-array-discluding-self/submission-1.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        const int size = nums.size();
        vector<int> new_nums(size, 1);
        int left = 1, right = 1, i = 0;

        for (i; i < size; i++) {
            new_nums[i] *= left;
            left *= nums[i];
        }

        for (i = size - 1; i >= 0; i--) {
            new_nums[i] *= right;
            right *= nums[i];
        }

        return new_nums;
    }
};
